Expose `get_reaction_force` and `get_reaction_torque` from 2D Joints (and `break_force` and `break_torque`)

Ughuuu opened the pull request:
- Fixes: https://github.com/godotengine/godot-proposals/issues/7672

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -196,6 +196,39 @@ real_t Joint2D::get_bias() const {
+		PhysicsServer2D::get_singleton()->joint_set_param(joint, PhysicsServer2D::JOINT_PARAM_BREAK_FORCE, break_force);
+	}
+}
+real_t Joint2D::get_break_force() const {
```
```suggestion

real_t Joint2D::get_break_force() const {
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -196,6 +196,39 @@ real_t Joint2D::get_bias() const {
 }
 
+void Joint2D::set_break_force(real_t p_break_force) {
+	break_force = p_break_force;
```
```suggestion
	if (break_force == p_break_force) {
		return;
	}
	break_force = p_break_force;
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -196,6 +196,39 @@ real_t Joint2D::get_bias() const {
+		PhysicsServer2D::get_singleton()->joint_set_param(joint, PhysicsServer2D::JOINT_PARAM_BREAK_TORQUE, break_torque);
+	}
+}
+real_t Joint2D::get_break_torque() const {
```
```suggestion

real_t Joint2D::get_break_torque() const {
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -196,6 +196,39 @@ real_t Joint2D::get_bias() const {
+}
+
+void Joint2D::set_break_torque(real_t p_break_torque) {
+	break_torque = p_break_torque;
```
```suggestion
	if (break_torque == p_break_torque) {
		return;
	}
	break_torque = p_break_torque;
```

AThousandShips commented on scene/2d/joint_2d.cpp:
```diff
@@ -196,6 +196,39 @@ real_t Joint2D::get_bias() const {
+	}
+	return 0.0;
+}
+real_t Joint2D::get_reaction_torque() const {
```
```suggestion

real_t Joint2D::get_reaction_torque() const {
```

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1073,11 +1079,32 @@ real_t GodotPhysicsServer2D::joint_get_param(RID p_joint, JointParam p_param) co
 
+real_t GodotPhysicsServer2D::joint_get_state(RID p_joint, JointState p_param) const {
+	const GodotJoint2D *joint = joint_owner.get_or_null(p_joint);
+	ERR_FAIL_COND_V(!joint, -1);
```
```suggestion
	ERR_FAIL_NULL_V(joint, -1);
```

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.h:
```diff
@@ -267,6 +267,8 @@ class GodotPhysicsServer2D : public PhysicsServer2D {
 
 	virtual void joint_set_param(RID p_joint, JointParam p_param, real_t p_value) override;
 	virtual real_t joint_get_param(RID p_joint, JointParam p_param) const override;
+	
```
Fix whitespace here

AThousandShips commented on doc/classes/Joint2D.xml:
```diff
@@ -9,6 +9,10 @@
 	<tutorials>
 	</tutorials>
 	<members>
+		<member name="break_force" type="float" setter="set_break_force" getter="get_break_force" default="0.0">
```
These should be further down, docs are sorted alphabetically

Ughuuu commented:
Is the API with get_state for joints good? Or should it be done in another way?

AThousandShips commented:
I think it looks proper, could change the `case` style to follow the generally established one but it doesn't follow it in the rest of the document, but that's one thing to maybe change later in the process once this has gotten some reviews and thoughts from the physics team

AThousandShips commented on servers/physics_server_2d.h:
```diff
@@ -828,6 +837,7 @@ VARIANT_ENUM_CAST(PhysicsServer2D::BodyDampMode);
 VARIANT_ENUM_CAST(PhysicsServer2D::BodyState);
 VARIANT_ENUM_CAST(PhysicsServer2D::CCDMode);
 VARIANT_ENUM_CAST(PhysicsServer2D::JointParam);
+VARIANT_ENUM_CAST(PhysicsServer2D::JointState);
```
You will need to register the values of of this enum in `_bind_methods` of `PhysicsServer2D`,

AThousandShips replied:
You need to bind them with `BIND_ENUM_CONSTANT` like you did in the other PR

Ughuuu commented:
Case style? Where exactly?

AThousandShips commented:
Putting brackets around, like so:
```cpp
case A: {
    foo();
} break;
```

But that's a style detail for later, but the only detail about that function I can think of

Ughuuu commented:
Will try to add an implementation soon(so far added the api part).

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1044,7 +1044,7 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
 void GodotPhysicsServer2D::joint_set_param(RID p_joint, JointParam p_param, real_t p_value) {
 	GodotJoint2D *joint = joint_owner.get_or_null(p_joint);
-	ERR_FAIL_COND(!joint);
+	ERR_FAIL_NULL(joint);
```
These are unnecessary, there's a PR for them coming soon so no need to change code outside of what you've already touched :)

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1044,40 +1044,67 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
+	switch (p_param) {
+		case JOINT_STATE_REACTION_FORCE:
+			return joint->get_reaction_force();
+			break;
+		case JOINT_STATE_REACTION_TORQUE:
+			return joint->get_reaction_torque();
+			break;
```
```suggestion
		case JOINT_STATE_REACTION_TORQUE:
			return joint->get_reaction_torque();
```
No need to add `break` when returning

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1044,40 +1044,67 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
 void GodotPhysicsServer2D::joint_set_param(RID p_joint, JointParam p_param, real_t p_value) {
 	GodotJoint2D *joint = joint_owner.get_or_null(p_joint);
-	ERR_FAIL_COND(!joint);
+	ERR_FAIL_COND(joint);
```
```suggestion
	ERR_FAIL_COND(!joint);
```
You forgot the `!` when restoring

AThousandShips commented on servers/physics_2d/godot_physics_server_2d.cpp:
```diff
@@ -1044,40 +1044,67 @@ void GodotPhysicsServer2D::joint_clear(RID p_joint) {
 real_t GodotPhysicsServer2D::joint_get_param(RID p_joint, JointParam p_param) const {
 	const GodotJoint2D *joint = joint_owner.get_or_null(p_joint);
-	ERR_FAIL_COND_V(!joint, -1);
+	ERR_FAIL_NULL_V(joint, -1);
```
```suggestion
	ERR_FAIL_COND_V(!joint, -1);
```

Ughuuu commented:
About disabling the joint if break force is met, should I:

1. set node_a and node_b to none.
2. set the joint to disabled mode for process. In this case, the node would need changing to stop processing when it's in disabled mode(currently the joint still works). Also, what about the children nodes if its disabled, would it be fair to disable children nodes of a joint if the joint breaks?
3. Add a state enabled that says the enabled state of the joint

Ughuuu commented:
Still need to separate the angular and linear velocity into force and torque, will also need to wait for https://github.com/godotengine/godot/pull/81610 to get approved as that changes the impulse and torque for pin joint

Ughuuu commented on servers/physics_2d/godot_joints_2d.h:
```diff
@@ -38,11 +38,18 @@ class GodotJoint2D : public GodotConstraint2D {
 	real_t bias = 0;
 	real_t max_bias = 3.40282e+38;
 	real_t max_force = 3.40282e+38;
+	real_t break_force = 3.40282e+38;
```
Hm, actually, not sure, maybe it should be in all cases max float value? It's the break force, and if a force is above this, the joint breaks.

AThousandShips replied:
My point was that the default in `Joint2D` is 0 right?

Ughuuu replied:
Yes. And for now I will go with option to enable breakable part, unless someone wants different

AThousandShips replied:
What I mean is:
When creating a new joint, if you don't set the value manually, and use the get method, it'll return `0`, but the actual value in the server will be `3.40282e+38`

Ughuuu commented on doc/classes/Joint2D.xml:
```diff
@@ -8,11 +8,31 @@
 			When set to [code]0[/code], the default value from [member ProjectSettings.physics/2d/solver/default_constraint_bias] is used.
 		</member>
+		<member name="break_force" type="float" setter="set_break_force" getter="get_break_force" default="0.0">
+			What force should the joint break under.
```
Good question. I would assume setting a high number? Not sure what would be a good way. Or if the value 0 or negative value is set, breaking is disabled? I think breaking would mean disabling the joint, but it's not possible yet I think to do that. Will create an issue to track that and create a fix before I merge this.

AThousandShips replied:
I think adding new functionality without allowing it to be disabled is not a good option here, setting it to some very large value, or an option would be good

Ughuuu replied:
Agree. A high number would be best. Even though an option and having the breaking_force 0 would be cleaner and have the intent more obvious.

Ughuuu committed:
Expose get_reaction_force and get_reaction_torque from 2D Joints (and break_force and break_torque)

remove extra flag

typo

fix

update typo

fix return type for is_break_enabled

fix docs

Update joint_2d.cpp

Update joint_2d.cpp

Update joint_2d.cpp

Update physics_server_2d_dummy.h

Update physics_server_2d_dummy.h

Ughuuu commented:
Sorry, made some mistakes when rebasing.

Ughuuu commented:
Any updates on this?

Ughuuu commented:
Rebased
